### ISSUE TITLE
CTC Example Problem

### ISSUE THREAD

**Issue by jiarenyf** [apache/incubator-mxnet#8347]
Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues.

If the issue is non-technical, feel free to present the information in what you believe is the best form.

## Description
(Brief description of the problem in no more than 2 sentences.)
I wonder in "https://github.com/apache/incubator-mxnet/blob/f9e25b47c0faec07e940fedb718df2f38dd925b4/example/ctc/lstm.py#L103", if we can delete the line `softmax_loss = mx.sym.MakeLoss(softmax_class)`, and use ` softmax_loss = mx.sym.BlockGrad(softmax_class)` direclty ?

That is, I don't know the role of the function "MakeLoss" in this line ?

## Environment info (Required)
pip install mxnet-cu80, version 0.11.0

```
What to do:
1. Download the diagnosis script from https://raw.githubusercontent.com/apache/incubator-mxnet/master/tools/diagnose.py
2. Run the script using `python diagnose.py` and paste its output here.

```

Package used (Python/R/Scala/Julia):
(I'm using ...)

For Scala user, please provide:
1. Java version: (`java -version`)
2. Maven version: (`mvn -version`)
3. Scala runtime if applicable: (`scala -version`)

For R user, please provide R `sessionInfo()`:

## Build info (Required if built from source)

Compiler (gcc/clang/mingw/visual studio):

MXNet commit hash:
(Paste the output of `git rev-parse HEAD` here.)

Build config:
(Paste the content of config.mk, or the build command.)

## Error Message:
(Paste the complete error message, including stack trace.)

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1.
2.

## What have you tried to solve it?

1.
2.

**Comment by jiarenyf** [comment]
@pluskid @thinxer

**Comment by jiarenyf** [comment]
……

**Comment by jiarenyf** [comment]
……

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by anirudh2290** [comment]
@szha Please add the label "Example" "Question"

**Comment by thomelane** [comment]
@jiarenyf Good question. As I understand, this function is written to return two things: the ctc loss (which is being optimised and used for back-propagation) and the softmax over the classes across time which can then be used for other purposes (but we don't want to optimize for this).

Since this `lstm_unroll` function returns a symbol, and ideally we'd want to call backwards on this resultant symbol, we need to make both components of the returned symbol losses (even if we don't really want to use the softmax part as a loss). Otherwise the symbol would expect to be passed a gradient when backward method is called (which we don't want).

Given we've made the both losses with `MakeLoss` we have to then 'disable' the gradient propagation for the softmax component, and this is done with `BlockGrad`. So to answer your question I think we need to keep these lines.

@ yajiedesign as the original author of this example, do you want to comment further?

Otherwise, @indhub this issue is good to close, thanks!
